fendermint/vm/resolver/src/ipld.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::{future::Future, time::Duration};
+
+use async_stm::{atomically, queues::TQueueLike};
+use ipc_ipld_resolver::Client as ResolverClient;
+use ipc_sdk::subnet_id::SubnetID;
+
+use crate::pool::{ResolveQueue, ResolveTask};
+
+/// The IPLD Resolver takes resolution tasks from the [ResolvePool] and
+/// uses the [ipc_ipld_resolver] to fetch the content from subnets.
+pub struct IpldResolver {
+    client: ResolverClient,
+    queue: ResolveQueue,
+    retry_delay: Duration,
+    own_subnet_id: SubnetID,
+}
+
+impl IpldResolver {
+    pub fn new(
+        client: ResolverClient,
+        queue: ResolveQueue,
+        retry_delay: Duration,
+        own_subnet_id: SubnetID,
+    ) -> Self {
+        Self {
+            client,
+            queue,
+            retry_delay,
+            own_subnet_id,
+        }
+    }
+
+    /// Start taking tasks from the resolver pool and resolving them using the IPLD Resolver.
+    pub async fn run(self) {
+        loop {
+            let (task, use_own_subnet) = atomically(|| {
+                let task = self.queue.read()?;
+                let use_own_subnet = task.use_own_subnet()?;
+                Ok((task, use_own_subnet))
+            })
+            .await;
+
+            start_resolve(
+                task,
+                self.client.clone(),
+                self.queue.clone(),
+                self.retry_delay,
+                if use_own_subnet {
+                    Some(self.own_subnet_id.clone())
+                } else {
+                    None
+                },
+            );
+        }
+    }
+}
+
+/// Run task resolution in the background, so as not to block items from other
+/// subnets being tried.
+fn start_resolve(
+    task: ResolveTask,
+    client: ResolverClient,
+    queue: ResolveQueue,
+    retry_delay: Duration,
+    own_subnet_id: Option<SubnetID>,
+) {
+    tokio::spawn(async move {
+        let from_theirs = client.resolve(task.cid(), task.subnet_id());
+        let from_own = own_subnet_id.map(|subnet_id| client.resolve(task.cid(), subnet_id));
+
+        let (theirs, own) = tokio::join!(from_theirs, future_opt(from_own));
+
+        let err = match (theirs, own) {
+            (Err(e), _) => {
+                tracing::error!(error = e.to_string(), "failed to submit resolution task");
+                // The service is no longer listening, we might as well stop taking new tasks from the queue.
+                // By not quitting we should see this error every time there is a new task, which is at least is a constant reminder.
+                return;
+            }
+            (Ok(Ok(())), _) | (_, Some(Ok(Ok(())))) => None,
+            (Ok(Err(e)), _) => Some(e),
```

Comment:
Thanks for asking, it's definitely not easy to see from this why this is correct. It was actually the compiler telling me that if I added more cases they would never be matched after this. 

Okay, so to answer your question: no, we don't need `from_theirs` and `from_own` to succeed at the same time, we just want one of them to go through, we just want the data ASAP. I could use `FuturesUnordered` and react as soon as one is ready, but then I wouldn't know if I should reschedule or wait for the other to succeed. It may be easier overall to do that, I'm not sure.

The way this Bitswap library works is that it deposits the date into the `BitswapStore`, so all we get back is either a `()` to indicate success, or an `Error` if it exhausted all options. So I'm thinking as long as I just see success from one of them, we are fine, the data arrived. 

If one is `Err`, the other is most likely `Err` as well, the `Service` died, so we are only interested in `Ok`:
* The first arm matches if `from_theirs` was successful, or if we tried `from_own` and _that_ was a success. 
* The second arm checks if `from_theirs` failed, and ignores the result of `from_own` (`None` or `Some` doesn't matter), because if `from_theirs` was successful, it would have matched the first arm already. 

So this is a not as fast as possible in terms of not retrying as quickly as it could, but it sends as few requests as possible by waiting until it knows that both failed before retrying, and it only retries once.